tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,594 @@
+    },
+    runEnvironment: {
+      type: Object,
+    },
+    metrics: {
+      type: Array,
+    },
+    activeBar:{
+      type: Object,
+      observer: '_activeBarChanged',
+    },
+    selectedChannel: {
+      type: Array,
+      notify: true,
+    },
+    selectedMetricIdx: {
+      type: Number,
+      value: 0,
+    },
+    selectedChannelId: {
+      type: Number,
+      value: 0,
+      observer: '_selectedChannelIdChanged',
+    },
+    _topoData: {
+      type: Object,
+      computed:
+          '_computeTopoData(data, runEnvironment, metrics, selectedMetricIdx)',
+    },
+    _linkData: {
+      type: Object,
+      computed: '_computeLinkData(data)',
+    },
+    _minChannelId: {
+      type: Number,
+      computed: '_computeMinChannelId(_linkData)',
+    },
+    _maxChannelId: {
+      type: Number,
+      computed: '_computeMaxChannelId(_linkData)',
+    },
+    _xDimension: {
+      type: Number,
+      computed: '_computeXDimension(runEnvironment)',
+    },
+    _yDimension: {
+      type: Number,
+      computed: '_computeYDimension(runEnvironment)',
+    },
+    _totalCoreCount: {
+      type: Number,
+      computed: '_computeTotalCoreCount(_xDimension, _yDimension)',
+    },
+    _tpuType: {
+      type: String,
+      computed: '_computeTpuType(runEnvironment)',
+    },
+    _hostXStride: {
+      type: Number,
+      computed: '_computeHostXStride(_tpuType)',
+    },
+    _hostGridWidth: {
+      type: Number,
+    },
+    _hostGridHeight: {
+      type: Number,
+    },
+    _nodeGridHeight: {
+      type: Number,
+    },
+    _nodeGridWidth: {
+      type: Number,
+    },
+    _gLink: {
+      type: Object,
+    },
+  },
+  observers: ['drawTopology(_topoData, runEnvironment)'],
+  /**
+   * Computes the topoData to be loaded into the topology graph.
+   */
+  _computeTopoData: function(
+      data: podviewer.proto.PodStatsMap | undefined,
+      runEnvironment: podviewer.proto.RunEnvironment | undefined,
+      metrics: Array<podviewer.proto.StackLayer>,
+      idx: number): Array<TopoData> {
+    if (!data || !runEnvironment || !runEnvironment.topology || !metrics ||
+      idx >= metrics.length || idx < 0) {
+      return;
+    }
+    const xdim = runEnvironment.topology.xDimension;
+    return Object.keys(data.podStatsPerCore).map((core) => {
+        const podStats = data.podStatsPerCore[core];
+        return {
+          xdim: podStats.chipId % xdim,
+          ydim: Math.floor(podStats.chipId / xdim),
+          nid: podStats.nodeId,
+          cid: podStats.chipId,
+          rid: data.coreIdToReplicaIdMap[core], // replica id.
+          host: podStats.hostName,
+          value: podStats[metrics[idx].key],
+          total: podStats.totalDurationUs,
+        };
+    });
+  },
+  /**
+   * Compute the data to be rendered as links.
+   */
+  _computeLinkData: function(
+      data: podviewer.proto.PodStatsMap): LinkData {
+    if (!data || !data.channelDb || data.channelDb.length == 0) return;
```

Comment:
Might be better to return an empty object.

Reply:
Done.